src/utils/random.ts

```diff
@@ -22,3 +23,26 @@ export function shuffle(array: any[], seed: string) {
 
   return array
 }
+
+export function calculateHash(str: string) {
+  const hash = crypto.createHash('sha256')
+  hash.update(str)
+  return hash.digest('hex')
+}
+
+export function deterministicShuffle<T>(
+  objectArray: T[],
+  identifyingProperty: keyof T,
+  pepper: string
+): T[] {
+  const map = new Map<string, T>()
+  objectArray.forEach((element) => {
+    const hash = calculateHash(`${element[identifyingProperty]}${pepper}`)
+    map.set(hash, element)
+  })
+
+  // sort the tokens array based on hash of the identifying property and user's address
+  return Array.from(map.entries())
+    .sort((a, b) => a[0].localeCompare(b[0]))
```

Comment:
nit: this [should be faster](https://stackoverflow.com/questions/14677060/400x-sorting-speedup-by-switching-a-localecompareb-to-ab-1ab10) and more correct as we don't want locale aware comparison here.
```suggestion
    .sort(([a], [b]) => a < b ? -1 : (a > b ? 1 : 0))
```